README.md
# WSL Hello sudo

"WSL Hello sudo" is a Linux PAM module and companion Windows CLI apps that realize `sudo` by
biometric login of [Windows Hello](https://www.microsoft.com/en-us/windows/windows-hello) on Windows Subsystem for Linux (WSL).  
This PAM module allows you to authenticate `sudo` via face recognition, fingerprint authentication, and of couse machine-local PIN.
It runs in both WSL and WSL 2.

Both the Linux PAM module and Windows CLI app are written in Rust.
Please use it at your own risk. There is no warranty.

![demo](https://github.com/nullpo-head/WSL-Hello-sudo/blob/master/demo.gif)

"WSL Hello sudo" actually does __not__ modify your `sudo` command at all. It is a Linux PAM module.  
PAM, _Plaggable Authentication Module_, is a UNIX's module system that provides user authentication mechanisms to applications such as `sudo` or `su`. "WSL Hello sudo" is such a PAM module that lets applications use Windows Hello.

## Installation and Configuration

### Installation

The installation process is very simple.  
Please download the latest release package from GitHub Release and unpack it.  
Run `install.sh` inside the directory, and follow the instruction of `install.sh`

```ShellSession
$ wget http://github.com/nullpo-head/WSL-Hello-sudo/releases/latest/download/release.tar.gz
$ tar xvf release.tar.gz
$ cd release
$ ./install.sh
```

Although you don't have to care about the detailed installation process,  
`install.sh` does following things.

1. Copy a small Windows CLI app that launches Windows Hello to `C:\Users\your_account\pam_wsl_hello` (default location)  
2. Install a PAM module to your WSL system.
3. Create config files in `/etc/pam_wsl_hello/`
4. Create a pam-configs entry in `/usr/share/pam-configs/` for automatic PAM configuration
5. Create `uninstall.sh`

### Configuration

"WSL Hello sudo" is not a fork of `sudo` but a PAM module. You have to adjust the PAM configuration to make it effective.

#### Automatic configuration
On Ubuntu, you can use `sudo pam-auth-update` to show a list of installed PAM authentication modules, and select the ones you want to use for authentication (which will also affect sudo etc.)

The install scripts will install the required configuration. If you're not using the install script, you can copy the pam-config file from the release tarball to `/usr/share/pam-configs/`.

#### Manual configuration

If for some reason you do not want to use automatic configuration, you can configure `/etc/pam.d/sudo` manually.
I strongly recommend to set password of root first so that you can switch to it by `su`, in case you make some typo in the config of `sudo`.  
Add `auth sufficient pam_wsl_hello.so` to the top line of your `/etc/pam.d/sudo` like the following example  

```
#%PAM-1.0

auth       sufficient pam_wsl_hello.so
session    required   pam_env.so readenv=1 user_readenv=0
session    required   pam_env.so readenv=1 envfile=/etc/default/locale user_readenv=0
@include common-auth
@include common-account
@include common-session-noninteractive
```
Even if you fail to authenticate via Windows Hello, `sudo` moves on to the regular password authentication by this setting with `sufficient`.  

Other applications that authenticate users such as `su` can also utilize Windows Hello by this module.  
Even so, __I strongly recommend you to make either `sudo` or `su` free from this module to prevent from being locked out__

## Troubleshooting

### Windows Hello window appears in background.

The Windows Hello dialog sometimes appears in background.
In some cases, it even fails to recognize your face with some weird error message.
It seems a bug of Windows API. In that case, restarting Windows a couple of times might solve the problem.

### "Windows Hello is not invoked! `sudo` just prompts password!"

Maybe some error is happening. Unfortunately, `sudo` suppresses error messages from PAM modules.  
To debug "WSL Hello sudo", make it effective for `su` instead of `sudo`. `su` shows error messages from PAM modules,
so you can see what is going on.

For your information, the setting for `su` will be like the example below.
I will show only relevant two lines.
```
auth       sufficient pam_rootok.so
auth       sufficient pam_wsl_hello.so  
```

## Build

Both the Linux PAM module and the Windows CLI apps of "WSL Hello sudo" are written in Rust.
So, only `cargo` is required to build it.

To build "WSL Hello sudo", make sure you're in WSL and then just run `make`.

```ShellSession
$ git clone https://github.com/nullpo-head/WSL-Hello-sudo.git
$ cd WSL-Hello-sudo
$ make
```

## Internals

Windows Hello maintains RSA key-pairs for each Windows user in its TPM hardware, and tells success of authentication by signing given contents by the private key.
To utilize its API, "WSL Hello sudo" contains small Windows CLI apps that return public key and signed signature of given content.
On the other hand, the PAM module of "WSL Hello sudo" remembers the public keys of each Windows user who corresponds to each Linux user.
So, the PAM module authenticates the given Linux user by the following process.

0. The PAM module is launched by `sudo` and receives a Linux user to be authenticated
1. The PAM module launches the companion Windows app and sends a random value via WSL's interop bridge
2. The companion Windows app invokes Windows Hello
3. Windows Hello makes a signature of the given input by the private key of the current Windows user
4. The companion Windows app returns the signature
5. The PAM module verifies the signature by the public key of the Windows user who corresponds to the given Linux user.


CHANGELOG.md
# [2.0.0](https://github.com/nullpo-head/WSL-Hello-sudo/compare/v1.4.0...v2.0.0) (2021-12-04)


### Features

* Rewrite Windows components in Rust ([#26](https://github.com/nullpo-head/WSL-Hello-sudo/issues/26)) ([d87ad12](https://github.com/nullpo-head/WSL-Hello-sudo/commit/d87ad1238e0f64e907f929b84f2bfcaec607e436))


### BREAKING CHANGES

* The file location and interface of the Windows component are updated, though it was not intended to be used by end users.

Co-authored-by: Takaya Saeki <abc.tkys+pub@gmail.com>



# [1.4.0](https://github.com/nullpo-head/WSL-Hello-sudo/compare/v1.3.2...v1.4.0) (2021-11-23)


### Features

* move hello to foreground once window exists ([#27](https://github.com/nullpo-head/WSL-Hello-sudo/issues/27)) ([b0b9f54](https://github.com/nullpo-head/WSL-Hello-sudo/commit/b0b9f543fdef61dc3cb77bb0fea523afcb62fa65))



## [1.3.2](https://github.com/nullpo-head/WSL-Hello-sudo/compare/v1.3.1...v1.3.2) (2021-11-22)


### Bug Fixes

* quote hello_winpath in uninstall.sh ([#28](https://github.com/nullpo-head/WSL-Hello-sudo/issues/28)) ([bd7e009](https://github.com/nullpo-head/WSL-Hello-sudo/commit/bd7e009beb15d7248c590e521eed3ecbaec16515))



## [1.3.1](https://github.com/nullpo-head/WSL-Hello-sudo/compare/v1.3.0...v1.3.1) (2021-08-29)


### Bug Fixes

* **install:** Fix installation failure that occured because .exe files did not have the execution permissions. ([5788c93](https://github.com/nullpo-head/WSL-Hello-sudo/commit/5788c93f265201134d5afe719b607a715a12e07d))



# [1.3.0](https://github.com/nullpo-head/WSL-Hello-sudo/compare/v1.2.0...v1.3.0) (2021-08-29)


### Bug Fixes

* Fix warning that pointer doesn't outlive temporary CString ([#23](https://github.com/nullpo-head/WSL-Hello-sudo/issues/23)) ([97289e5](https://github.com/nullpo-head/WSL-Hello-sudo/commit/97289e55b2e85e4e5994f0e8b9617900e3ae8079))
* quote PAM_WSL_HELLO_WINPATH to allow spaces ([abad2f8](https://github.com/nullpo-head/WSL-Hello-sudo/commit/abad2f8c79e02d1436899a1a89af67b2efd7b08d))


### Features

* add pam-config for automatic configuration Install automatic PAM config ([1380e27](https://github.com/nullpo-head/WSL-Hello-sudo/commit/1380e2749f79bf0dd0255ec7a7ebc360b8b71290)), closes [#14](https://github.com/nullpo-head/WSL-Hello-sudo/issues/14)
* support custom /c mountpoints ([#13](https://github.com/nullpo-head/WSL-Hello-sudo/issues/13)) ([5fcc2a1](https://github.com/nullpo-head/WSL-Hello-sudo/commit/5fcc2a183956f5db27359b3a798d3b77a8e14e8b))


### Performance Improvements

* decrease pam module binary size ([c79a184](https://github.com/nullpo-head/WSL-Hello-sudo/commit/c79a18458ac71d11d3e99af5eecdb4adb17d34f0))



# [1.2.0](https://github.com/nullpo-head/WSL-Hello-sudo/compare/v1.1.0...v1.2.0) (2020-05-29)



# [1.1.0](https://github.com/nullpo-head/WSL-Hello-sudo/compare/v1.0.0...v1.1.0) (2020-05-29)



# 1.0.0 (2017-09-18)





wsl_hello_pam/src/bindings.rs
#![allow(clippy::all)]
/* Based on bindings automatically generated by rust-bindgen */

pub const PAM_SUCCESS: ::std::os::raw::c_int = 0;
pub const PAM_OPEN_ERR: ::std::os::raw::c_int = 1;
pub const PAM_SYMBOL_ERR: ::std::os::raw::c_int = 2;
pub const PAM_SERVICE_ERR: ::std::os::raw::c_int = 3;
pub const PAM_SYSTEM_ERR: ::std::os::raw::c_int = 4;
pub const PAM_BUF_ERR: ::std::os::raw::c_int = 5;
pub const PAM_PERM_DENIED: ::std::os::raw::c_int = 6;
pub const PAM_AUTH_ERR: ::std::os::raw::c_int = 7;
pub const PAM_CRED_INSUFFICIENT: ::std::os::raw::c_int = 8;
pub const PAM_AUTHINFO_UNAVAIL: ::std::os::raw::c_int = 9;
pub const PAM_USER_UNKNOWN: ::std::os::raw::c_int = 10;
pub const PAM_MAXTRIES: ::std::os::raw::c_int = 11;
pub const PAM_NEW_AUTHTOK_REQD: ::std::os::raw::c_int = 12;
pub const PAM_ACCT_EXPIRED: ::std::os::raw::c_int = 13;
pub const PAM_SESSION_ERR: ::std::os::raw::c_int = 14;
pub const PAM_CRED_UNAVAIL: ::std::os::raw::c_int = 15;
pub const PAM_CRED_EXPIRED: ::std::os::raw::c_int = 16;
pub const PAM_CRED_ERR: ::std::os::raw::c_int = 17;
pub const PAM_NO_MODULE_DATA: ::std::os::raw::c_int = 18;
pub const PAM_CONV_ERR: ::std::os::raw::c_int = 19;
pub const PAM_AUTHTOK_ERR: ::std::os::raw::c_int = 20;
pub const PAM_AUTHTOK_RECOVERY_ERR: ::std::os::raw::c_int = 21;
pub const PAM_AUTHTOK_LOCK_BUSY: ::std::os::raw::c_int = 22;
pub const PAM_AUTHTOK_DISABLE_AGING: ::std::os::raw::c_int = 23;
pub const PAM_TRY_AGAIN: ::std::os::raw::c_int = 24;
pub const PAM_IGNORE: ::std::os::raw::c_int = 25;
pub const PAM_ABORT: ::std::os::raw::c_int = 26;
pub const PAM_AUTHTOK_EXPIRED: ::std::os::raw::c_int = 27;
pub const PAM_MODULE_UNKNOWN: ::std::os::raw::c_int = 28;
pub const PAM_BAD_ITEM: ::std::os::raw::c_int = 29;
pub const PAM_CONV_AGAIN: ::std::os::raw::c_int = 30;
pub const PAM_INCOMPLETE: ::std::os::raw::c_int = 31;
pub const PAM_SILENT: ::std::os::raw::c_int = 32768;
pub const PAM_DISALLOW_NULL_AUTHTOK: ::std::os::raw::c_int = 1;
pub const PAM_ESTABLISH_CRED: ::std::os::raw::c_int = 2;
pub const PAM_DELETE_CRED: ::std::os::raw::c_int = 4;
pub const PAM_REINITIALIZE_CRED: ::std::os::raw::c_int = 8;
pub const PAM_REFRESH_CRED: ::std::os::raw::c_int = 16;
pub const PAM_CHANGE_EXPIRED_AUTHTOK: ::std::os::raw::c_int = 32;
pub const PAM_SERVICE: ::std::os::raw::c_int = 1;
pub const PAM_USER: ::std::os::raw::c_int = 2;
pub const PAM_TTY: ::std::os::raw::c_int = 3;
pub const PAM_RHOST: ::std::os::raw::c_int = 4;
pub const PAM_CONV: ::std::os::raw::c_int = 5;
pub const PAM_AUTHTOK: ::std::os::raw::c_int = 6;
pub const PAM_OLDAUTHTOK: ::std::os::raw::c_int = 7;
pub const PAM_RUSER: ::std::os::raw::c_int = 8;
pub const PAM_USER_PROMPT: ::std::os::raw::c_int = 9;
pub const PAM_FAIL_DELAY: ::std::os::raw::c_int = 10;
pub const PAM_XDISPLAY: ::std::os::raw::c_int = 11;
pub const PAM_XAUTHDATA: ::std::os::raw::c_int = 12;
pub const PAM_AUTHTOK_TYPE: ::std::os::raw::c_int = 13;
pub const PAM_DATA_SILENT: ::std::os::raw::c_int = 1073741824;
pub const PAM_PROMPT_ECHO_OFF: ::std::os::raw::c_int = 1;
pub const PAM_PROMPT_ECHO_ON: ::std::os::raw::c_int = 2;
pub const PAM_ERROR_MSG: ::std::os::raw::c_int = 3;
pub const PAM_TEXT_INFO: ::std::os::raw::c_int = 4;
pub const PAM_RADIO_TYPE: ::std::os::raw::c_int = 5;
pub const PAM_BINARY_PROMPT: ::std::os::raw::c_int = 7;
pub const PAM_MAX_NUM_MSG: ::std::os::raw::c_int = 32;
pub const PAM_MAX_MSG_SIZE: ::std::os::raw::c_int = 512;
pub const PAM_MAX_RESP_SIZE: ::std::os::raw::c_int = 512;
pub const PAM_AUTHTOK_RECOVER_ERR: ::std::os::raw::c_int = 21;
pub const PAM_PRELIM_CHECK: ::std::os::raw::c_int = 16384;
pub const PAM_UPDATE_AUTHTOK: ::std::os::raw::c_int = 8192;
pub const PAM_DATA_REPLACE: ::std::os::raw::c_int = 536870912;

#[repr(C)]
#[derive(Debug, Copy, Clone)]
pub struct pam_handle {
    _unused: [u8; 0],
}
pub type pam_handle_t = pam_handle;
extern "C" {
    pub fn pam_set_item(
        pamh: *mut pam_handle_t,
        item_type: ::std::os::raw::c_int,
        item: *const ::std::os::raw::c_void,
    ) -> ::std::os::raw::c_int;
}
extern "C" {
    pub fn pam_get_item(
        pamh: *const pam_handle_t,
        item_type: ::std::os::raw::c_int,
        item: *mut *const ::std::os::raw::c_void,
    ) -> ::std::os::raw::c_int;
}
extern "C" {
    pub fn pam_strerror(
        pamh: *mut pam_handle_t,
        errnum: ::std::os::raw::c_int,
    ) -> *const ::std::os::raw::c_char;
}
extern "C" {
    pub fn pam_putenv(
        pamh: *mut pam_handle_t,
        name_value: *const ::std::os::raw::c_char,
    ) -> ::std::os::raw::c_int;
}
extern "C" {
    pub fn pam_getenv(
        pamh: *mut pam_handle_t,
        name: *const ::std::os::raw::c_char,
    ) -> *const ::std::os::raw::c_char;
}
extern "C" {
    pub fn pam_getenvlist(pamh: *mut pam_handle_t) -> *mut *mut ::std::os::raw::c_char;
}
extern "C" {
    pub fn pam_fail_delay(
        pamh: *mut pam_handle_t,
        musec_delay: ::std::os::raw::c_uint,
    ) -> ::std::os::raw::c_int;
}
#[repr(C)]
#[derive(Debug, Copy)]
pub struct pam_message {
    pub msg_style: ::std::os::raw::c_int,
    pub msg: *const ::std::os::raw::c_char,
}
#[test]
fn bindgen_test_layout_pam_message() {
    assert_eq!(
        ::std::mem::size_of::<pam_message>(),
        16usize,
        concat!("Size of: ", stringify!(pam_message))
    );
    assert_eq!(
        ::std::mem::align_of::<pam_message>(),
        8usize,
        concat!("Alignment of ", stringify!(pam_message))
    );
    assert_eq!(
        unsafe { &(*(0 as *const pam_message)).msg_style as *const _ as usize },
        0usize,
        concat!(
            "Alignment of field: ",
            stringify!(pam_message),
            "::",
            stringify!(msg_style)
        )
    );
    assert_eq!(
        unsafe { &(*(0 as *const pam_message)).msg as *const _ as usize },
        8usize,
        concat!(
            "Alignment of field: ",
            stringify!(pam_message),
            "::",
            stringify!(msg)
        )
    );
}
impl Clone for pam_message {
    fn clone(&self) -> Self {
        *self
    }
}
#[repr(C)]
#[derive(Debug, Copy)]
pub struct pam_response {
    pub resp: *mut ::std::os::raw::c_char,
    pub resp_retcode: ::std::os::raw::c_int,
}
#[test]
fn bindgen_test_layout_pam_response() {
    assert_eq!(
        ::std::mem::size_of::<pam_response>(),
        16usize,
        concat!("Size of: ", stringify!(pam_response))
    );
    assert_eq!(
        ::std::mem::align_of::<pam_response>(),
        8usize,
        concat!("Alignment of ", stringify!(pam_response))
    );
    assert_eq!(
        unsafe { &(*(0 as *const pam_response)).resp as *const _ as usize },
        0usize,
        concat!(
            "Alignment of field: ",
            stringify!(pam_response),
            "::",
            stringify!(resp)
        )
    );
    assert_eq!(
        unsafe { &(*(0 as *const pam_response)).resp_retcode as *const _ as usize },
        8usize,
        concat!(
            "Alignment of field: ",
            stringify!(pam_response),
            "::",
            stringify!(resp_retcode)
        )
    );
}
impl Clone for pam_response {
    fn clone(&self) -> Self {
        *self
    }
}
#[repr(C)]
#[derive(Debug, Copy)]
pub struct pam_conv {
    pub conv: ::std::option::Option<
        unsafe extern "C" fn(
            num_msg: ::std::os::raw::c_int,
            msg: *mut *const pam_message,
            resp: *mut *mut pam_response,
            appdata_ptr: *mut ::std::os::raw::c_void,
        ) -> ::std::os::raw::c_int,
    >,
    pub appdata_ptr: *mut ::std::os::raw::c_void,
}
#[test]
fn bindgen_test_layout_pam_conv() {
    assert_eq!(
        ::std::mem::size_of::<pam_conv>(),
        16usize,
        concat!("Size of: ", stringify!(pam_conv))
    );
    assert_eq!(
        ::std::mem::align_of::<pam_conv>(),
        8usize,
        concat!("Alignment of ", stringify!(pam_conv))
    );
    assert_eq!(
        unsafe { &(*(0 as *const pam_conv)).conv as *const _ as usize },
        0usize,
        concat!(
            "Alignment of field: ",
            stringify!(pam_conv),
            "::",
            stringify!(conv)
        )
    );
    assert_eq!(
        unsafe { &(*(0 as *const pam_conv)).appdata_ptr as *const _ as usize },
        8usize,
        concat!(
            "Alignment of field: ",
            stringify!(pam_conv),
            "::",
            stringify!(appdata_ptr)
        )
    );
}
impl Clone for pam_conv {
    fn clone(&self) -> Self {
        *self
    }
}

#[repr(C)]
#[derive(Debug, Copy)]
pub struct pam_xauth_data {
    pub namelen: ::std::os::raw::c_int,
    pub name: *mut ::std::os::raw::c_char,
    pub datalen: ::std::os::raw::c_int,
    pub data: *mut ::std::os::raw::c_char,
}

#[test]
fn bindgen_test_layout_pam_xauth_data() {
    assert_eq!(
        ::std::mem::size_of::<pam_xauth_data>(),
        32usize,
        concat!("Size of: ", stringify!(pam_xauth_data))
    );
    assert_eq!(
        ::std::mem::align_of::<pam_xauth_data>(),
        8usize,
        concat!("Alignment of ", stringify!(pam_xauth_data))
    );
    assert_eq!(
        unsafe { &(*(0 as *const pam_xauth_data)).namelen as *const _ as usize },
        0usize,
        concat!(
            "Alignment of field: ",
            stringify!(pam_xauth_data),
            "::",
            stringify!(namelen)
        )
    );
    assert_eq!(
        unsafe { &(*(0 as *const pam_xauth_data)).name as *const _ as usize },
        8usize,
        concat!(
            "Alignment of field: ",
            stringify!(pam_xauth_data),
            "::",
            stringify!(name)
        )
    );
    assert_eq!(
        unsafe { &(*(0 as *const pam_xauth_data)).datalen as *const _ as usize },
        16usize,
        concat!(
            "Alignment of field: ",
            stringify!(pam_xauth_data),
            "::",
            stringify!(datalen)
        )
    );
    assert_eq!(
        unsafe { &(*(0 as *const pam_xauth_data)).data as *const _ as usize },
        24usize,
        concat!(
            "Alignment of field: ",
            stringify!(pam_xauth_data),
            "::",
            stringify!(data)
        )
    );
}
impl Clone for pam_xauth_data {
    fn clone(&self) -> Self {
        *self
    }
}

#[link(name = "pam")]
extern "C" {
    pub fn pam_start(
        service_name: *const ::std::os::raw::c_char,
        user: *const ::std::os::raw::c_char,
        pam_conversation: *const pam_conv,
        pamh: *mut *mut pam_handle_t,
    ) -> ::std::os::raw::c_int;

    pub fn pam_end(
        pamh: *mut pam_handle_t,
        pam_status: ::std::os::raw::c_int,
    ) -> ::std::os::raw::c_int;

    pub fn pam_authenticate(
        pamh: *mut pam_handle_t,
        flags: ::std::os::raw::c_int,
    ) -> ::std::os::raw::c_int;

    pub fn pam_setcred(
        pamh: *mut pam_handle_t,
        flags: ::std::os::raw::c_int,
    ) -> ::std::os::raw::c_int;

    pub fn pam_acct_mgmt(
        pamh: *mut pam_handle_t,
        flags: ::std::os::raw::c_int,
    ) -> ::std::os::raw::c_int;

    pub fn pam_open_session(
        pamh: *mut pam_handle_t,
        flags: ::std::os::raw::c_int,
    ) -> ::std::os::raw::c_int;

    pub fn pam_close_session(
        pamh: *mut pam_handle_t,
        flags: ::std::os::raw::c_int,
    ) -> ::std::os::raw::c_int;

    pub fn pam_chauthtok(
        pamh: *mut pam_handle_t,
        flags: ::std::os::raw::c_int,
    ) -> ::std::os::raw::c_int;

    pub fn pam_set_data(
        pamh: *mut pam_handle_t,
        module_data_name: *const ::std::os::raw::c_char,
        data: *mut ::std::os::raw::c_void,
        cleanup: ::std::option::Option<
            unsafe extern "C" fn(
                pamh: *mut pam_handle_t,
                data: *mut ::std::os::raw::c_void,
                error_status: ::std::os::raw::c_int,
            ),
        >,
    ) -> ::std::os::raw::c_int;

    pub fn pam_get_data(
        pamh: *const pam_handle_t,
        module_data_name: *const ::std::os::raw::c_char,
        data: *mut *const ::std::os::raw::c_void,
    ) -> ::std::os::raw::c_int;

    pub fn pam_get_user(
        pamh: *mut pam_handle_t,
        user: *mut *const ::std::os::raw::c_char,
        prompt: *const ::std::os::raw::c_char,
    ) -> ::std::os::raw::c_int;
}

/*
extern "C" {
    pub fn pam_sm_authenticate(pamh: *mut pam_handle_t,
                               flags: ::std::os::raw::c_int,
                               argc: ::std::os::raw::c_int,
                               argv: *mut *const ::std::os::raw::c_char)
     -> ::std::os::raw::c_int;
}
extern "C" {
    pub fn pam_sm_setcred(pamh: *mut pam_handle_t,
                          flags: ::std::os::raw::c_int,
                          argc: ::std::os::raw::c_int,
                          argv: *mut *const ::std::os::raw::c_char)
     -> ::std::os::raw::c_int;
}
extern "C" {
    pub fn pam_sm_acct_mgmt(pamh: *mut pam_handle_t,
                            flags: ::std::os::raw::c_int,
                            argc: ::std::os::raw::c_int,
                            argv: *mut *const ::std::os::raw::c_char)
     -> ::std::os::raw::c_int;
}
extern "C" {
    pub fn pam_sm_open_session(pamh: *mut pam_handle_t,
                               flags: ::std::os::raw::c_int,
                               argc: ::std::os::raw::c_int,
                               argv: *mut *const ::std::os::raw::c_char)
     -> ::std::os::raw::c_int;
}
extern "C" {
    pub fn pam_sm_close_session(pamh: *mut pam_handle_t,
                                flags: ::std::os::raw::c_int,
                                argc: ::std::os::raw::c_int,
                                argv: *mut *const ::std::os::raw::c_char)
     -> ::std::os::raw::c_int;
}
extern "C" {
    pub fn pam_sm_chauthtok(pamh: *mut pam_handle_t,
                            flags: ::std::os::raw::c_int,
                            argc: ::std::os::raw::c_int,
                            argv: *mut *const ::std::os::raw::c_char)
     -> ::std::os::raw::c_int;
}
*/
